modules/image-loading-optimization/module.php

```diff
@@ -0,0 +1,346 @@
+
+		return $optimized_image;
+	}
+
+	/**
+	 * Adds optimization attributes to an `img` HTML tag.
+	 *
+	 * @param string $image   The HTML `img` tag where the attribute should be added.
+	 * @return string Converted `img` tag with optimization attributes added.
+	 */
+	private function add_loading_optimization_attrs( $image ) {
+		$width             = preg_match( '/ width=["\']([0-9]+)["\']/', $image, $match_width ) ? (int) $match_width[1] : null;
+		$height            = preg_match( '/ height=["\']([0-9]+)["\']/', $image, $match_height ) ? (int) $match_height[1] : null;
+		$loading_val       = preg_match( '/ loading=["\']([A-Za-z]+)["\']/', $image, $match_loading ) ? $match_loading[1] : null;
+		$fetchpriority_val = preg_match( '/ fetchpriority=["\']([A-Za-z]+)["\']/', $image, $match_fetchpriority ) ? $match_fetchpriority[1] : null;
+
+		// Images should have height and dimension width for the loading optimization attributes to be added.
+		if ( ! str_contains( $image, ' width="' ) || ! str_contains( $image, ' height="' ) ) {
+			return $image;
+		}
+
+		$optimization_attrs = $this->get_loading_optimization_attributes(
+			array(
+				'width'         => $width,
+				'height'        => $height,
+				'loading'       => $loading_val,
+				'fetchpriority' => $fetchpriority_val,
+			)
+		);
+
+		if ( ! empty( $optimization_attrs['fetchpriority'] ) ) {
+			$image = str_replace( '<img', '<img fetchpriority="' . esc_attr( $optimization_attrs['fetchpriority'] ) . '"', $image );
+		}
+
+		if ( ! empty( $optimization_attrs['loading'] ) ) {
+			$image = str_replace( '<img', '<img loading="' . esc_attr( $optimization_attrs['loading'] ) . '"', $image );
+		}
+
+		return $image;
+	}
+
+	/**
+	 * Return loading Loading optimization attributes for a image with give attribute.
+	 *
+	 * @param array $attr Existing image attributes.
+	 * @return array Loading optimization attributes.
+	 */
+	private function get_loading_optimization_attributes( $attr ) {
+		$loading_attrs = array();
+
+		// For any resources, width and height must be provided, to avoid layout shifts.
+		if ( ! isset( $attr['width'], $attr['height'] ) ) {
+			return $loading_attrs;
+		}
+
+		/*
+		 * The key function logic starts here.
+		 */
+		$maybe_in_viewport    = null;
+		$increase_count       = false;
+		$maybe_increase_count = false;
+
+		// Logic to handle a `loading` attribute that is already provided.
+		if ( isset( $attr['loading'] ) ) {
+			/*
+			 * Interpret "lazy" as not in viewport. Any other value can be
+			 * interpreted as in viewport (realistically only "eager" or `false`
+			 * to force-omit the attribute are other potential values).
+			 */
+			if ( 'lazy' === $attr['loading'] ) {
+				$maybe_in_viewport = false;
+			} else {
+				$maybe_in_viewport = true;
+			}
+		}
+
+		// Logic to handle a `fetchpriority` attribute that is already provided.
+		if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
```

Comment:
maybe extract to with is_high_fetch_priority() -> make it more readable

Reply:
See my question above. @TzviRabinovitch